### DISCUSSION TITLE
Expand letter spacing to fill width

### DISCUSSION THREAD

**Discussion by N4tus** [typst/typst#4487] [Q&A] [answered]
I have a text-line, a title, and I want to expand the letter-spacing such that the title fills the width of its parent container.
Currently I manually adjust the `tracking` property of `text` to achieve this.

**Comment by jamesrswift** [comment]
A solution such as this might be of use (but change text size to tracking). Credit to Enivex

```
#set align(center)


#let text-stretch(body, width: 100pt) = context {
  let scale-factor = measure(text(size: 10pt,body)).width / width.to-absolute()
  text(size: 10pt / scale-factor, body)
}

#let text-stack(..input) = stack(
  spacing: 0.5em,
  ..(
    for entries in input.pos() {
      (text-stretch(entries, ..input.named()),)
    }
  )
)

#text-stack(width: 1.5in,
  "PHILOSOPHIÆ", "NATURALIS", "PRINCIPIA", "MATHEMATICA",
)
```

**Comment by Enivex** [comment]
> A solution such as this might be of use (but change text size to tracking). Credit to Enivex

You can do a variation of this, but for this particular problem I think you would need to measure two data points in order to get the right value. (Because zero tracking doesn't mean zero width.)

**Comment by jamesrswift** [comment]
Indeed, I'd recommend measuring at 0pt kerning and at 1pt kerning, which
should give the gradient to figure out which kerning value to finish with

**Comment by jamesrswift** [comment, marked as answer]
This should do the trick:
```typ
#let text-stretch(body, width: 100pt) = context {
  let x1 = measure(text(tracking: 0pt,body)).width
  let x2 = measure(text(tracking: 1pt,body)).width
  let tracking = (width - x1)/(x2 - x1)
  text(tracking: tracking * 1pt, body)
}
```
yields
![space to fill](https://github.com/typst/typst/assets/6299280/5e2e194d-533c-4934-8214-4b4c1e31e784)

**Reply by N4tus** [reply]
This looks really good. Only thing worth noting for future readers is, that you might want to turn off `ligatures`.
